compiler/frontend/src/org/jetbrains/kotlin/types/expressions/ExpressionTypingVisitorForStatements.java

```diff
@@ -441,6 +443,36 @@ protected KotlinTypeInfo visitAssignment(KtBinaryExpression expression, Expressi
                                   ? refineTypeByPropertyInType(bindingContext, leftOperand, leftType)
                                   : refineTypeFromPropertySetterIfPossible(bindingContext, leftOperand, leftType);
 
+
+        // Resolve assign overload
+        VariableDescriptor descriptor = BindingContextUtils.extractVariableFromResolvedCall(bindingContext, leftOperand);
+        OverloadResolutionResults<FunctionDescriptor> assignmentOperationDescriptors = new NameNotFoundResolutionResult<>();
+        KotlinType assignmentOperationType = null;
+        TemporaryTraceAndCache temporaryForAssignmentOperation = null;
+        boolean isDelegated = descriptor instanceof PropertyDescriptor && ((PropertyDescriptor) descriptor).isDelegated();
+        if (descriptor != null && !descriptor.isVar() && !isDelegated) {
+            ExpressionReceiver receiver = ExpressionReceiver.Companion.create(left, leftType, context.trace.getBindingContext());
+            temporaryForAssignmentOperation = TemporaryTraceAndCache.create(
+                    context, "trace to check assignment operation like '=' for", expression);
+            assignmentOperationDescriptors =
+                    components.callResolver.resolveBinaryCall(
+                            context.replaceTraceAndCache(temporaryForAssignmentOperation).replaceScope(scope),
+                            receiver, expression, ASSIGN
+                    );
+            assignmentOperationType = OverloadResolutionResultsUtil.getResultingType(assignmentOperationDescriptors, context);
+        }
+        boolean isResolvedAssignOverload = assignmentOperationType != null && assignmentOperationDescriptors.isSuccess();
+        if (isResolvedAssignOverload) {
+            temporaryForAssignmentOperation.commit();
+            if (!KotlinTypeChecker.DEFAULT.equalTypes(components.builtIns.getUnitType(), assignmentOperationType)) {
+                KtSimpleNameExpression operationSign = expression.getOperationReference();
+                context.trace.report(ASSIGNMENT_OPERATOR_SHOULD_RETURN_UNIT.on(operationSign, assignmentOperationDescriptors.getResultingDescriptor(), operationSign));
```

Comment:
Why do you report this diagnostic on call site instead of on declaration of the `assign` fun?

Reply:
For plusAssign&friends there is the same logic. Do you maybe know the reason why there is that kind of logic?

Reply:
Oh, I get it. This is needed for a case when `plusAssign` came from java class, so we don't have a declaration site to report an error. So it's fine to leave it here